### PR TITLE
Implode voices to voices

### PR DESCRIPTION
**Pull request by Rockettwo**
Added a basic implementation to implode from voices to voices. Actually we can only merge four voices at once, so in fact we are filling just one stave. The source staffs don't change only the destination staff is changed. The implementation always uses full measures but in the next attempts this will get fixed. 
There is a pre-check for empty voices, so we're just copying existing voices which don't have just rests. Actually everything which isn't part of a ChordRest is ignored, so there are no spanners, dynamics, glissandi and so on. But this is also the current implementation. There is a discussion [here](https://musescore.org/en/node/107781) about an extended Implode tool but for this there are quite more things we have to check.
I didn't update the test yet ;)
EDIT: tests updated now. More tests coming

### CODE REVIEW

File: libmscore/cmd.cpp [hunk 1]
```diff
@@ -2680,88 +2680,101 @@ void Score::cmdImplode()
...
-                                          foreach (Note *tn, tied->notes()) {
-                                                if (nn->pitch() == tn->pitch() && nn->tpc() == tn->tpc() && !tn->tieFor()) {
-                                                      // found note to tie
-                                                      Tie* tie = new Tie(this);
-                                                      tie->setStartNote(tn);
-                                                      tie->setEndNote(nn);
-                                                      tie->setTrack(tn->track());
-                                                      undoAddElement(tie);
+                                    // add notes
+                                    foreach (Note* n, srcChord->notes()) {
```
Comment (Jojo-Schmitz) [inline libmscore/cmd.cpp]: please convert to `for(Note* n : tied()->notes())`

File: libmscore/cmd.cpp [hunk 2]
```diff
@@ -2680,88 +2680,101 @@ void Score::cmdImplode()
...
+                                          NoteVal nv(n->pitch());
+                                          nv.tpc1 = n->tpc1();
+                                          // skip duplicates
+                                          if (dstChord->findNote(nv.pitch))
+                                                continue;
+                                          Note* nn = addNote(dstChord, nv);
+                                          // add tie to this note if original chord was tied
+                                          if (tied) {
+                                                // find note to tie to
+                                                foreach (Note *tn, tied->notes()) {
```
Comment (Jojo-Schmitz) [inline libmscore/cmd.cpp]: please convert to `for(Note* tn : tied()->notes())`

File: libmscore/edit.cpp [hunk 3]
```diff
@@ -3075,4 +3075,176 @@ void Score::checkSpanner(int startTick, int endTick)
...
+                              nt->setTrack(dtrack);
+                              nt->setParent(dm);
+                              tupletMap.add(ot, nt);
+
+                              Tuplet* nt1 = nt;
+                              while (ot->tuplet()) {
+                                    Tuplet* nt = tupletMap.findNew(ot->tuplet());
+                                    if (nt == 0) {
+                                          if (link)
+                                                nt = static_cast<Tuplet*>(ot->tuplet()->linkedClone());
```
Comment (Jojo-Schmitz) [inline libmscore/edit.cpp]: please `toTuplet(...)`

File: libmscore/edit.cpp [hunk 4]
```diff
@@ -3075,4 +3075,176 @@ void Score::checkSpanner(int startTick, int endTick)
...
+                                    nt1 = nt;
+                                    }
+                              }
+                        nt->add(ncr);
+                        ncr->setTuplet(nt);
+                        }
+
+                  if (oe->isChordRest()) {
+                        if (oe->isChord()) {
+                              Chord* och = static_cast<Chord*>(ocr);
```
Comment (Jojo-Schmitz) [inline libmscore/edit.cpp]: `toChord(...)`

File: libmscore/edit.cpp [hunk 5]
```diff
@@ -3075,4 +3075,176 @@ void Score::checkSpanner(int startTick, int endTick)
...
+                                                }
+                                          }
+                                    }
+                              // two note tremolo
+                              if (och->tremolo() && och->tremolo()->twoNotes()) {
+                                   if (och == och->tremolo()->chord1()) {
+                                          if (tremolo)
+                                                qDebug("unconnected two note tremolo");
+                                          if (link)
+                                                tremolo = static_cast<Tremolo*>(och->tremolo()->linkedClone());
```
Comment (Jojo-Schmitz) [inline libmscore/edit.cpp]: `toTremolo(...)`

File: libmscore/edit.cpp [hunk 6]
```diff
@@ -3075,4 +3075,176 @@ void Score::checkSpanner(int startTick, int endTick)
...
+                              Chord* nch = static_cast<Chord*>(ncr);
+
+                              int n = och->notes().size();
+                              for (int i = 0; i < n; ++i) {
+                                    Note* on = och->notes().at(i);
+                                    Note* nn = nch->notes().at(i);
+                                    if (on->tieFor()) {
+                                          Tie* tie;
+                                          if (link)
+                                                tie = static_cast<Tie*>(on->tieFor()->linkedClone());
```
Comment (Jojo-Schmitz) [inline libmscore/edit.cpp]: `toTie()`

File: libmscore/cmd.cpp [hunk 7]
```diff
@@ -2680,88 +2680,101 @@ void Score::cmdImplode()
...
+            // loop through segments adding notes to chord on top staff
+            for (Segment* s = startSegment; s && s != endSegment; s = s->next1()) {
+                  if (s->segmentType() != Segment::Type::ChordRest)
+                        continue;
+                  Element* dst = s->element(dstTrack);
+                  if (dst && dst->type() == Element::Type::CHORD) {
+                        Chord* dstChord = toChord(dst);
+                        // see if we are tying in to this chord
+                        Chord* tied = 0;
+                        foreach (Note* n, dstChord->notes()) {
```
Comment (Jojo-Schmitz) [inline libmscore/cmd.cpp]: `for (Note* n : dstChord->notes())`

File: libmscore/cmd.cpp [hunk 8]
```diff
@@ -2680,88 +2680,101 @@ void Score::cmdImplode()
...
-                        if (n->tieBack()) {
-                              tied = n->tieBack()->startNote()->chord();
-                              break;
+      // if single staff selected, combine voices
+      // otherwise combine staves
+      if (dstStaff == endStaff - 1) {
+
+            // loop through segments adding notes to chord on top staff
+            for (Segment* s = startSegment; s && s != endSegment; s = s->next1()) {
+                  if (s->segmentType() != Segment::Type::ChordRest)
```
Comment (Jojo-Schmitz) [inline libmscore/cmd.cpp]: `s->segmentType.isChordRest()`

File: libmscore/edit.cpp (hunk 6 above)
Comment (Rockettwo) [inline libmscore/edit.cpp]: Things in edit.cpp are wrong in my other PR, too then but if we assume this one is merged after the other...

**Comment by lasconic** [comment]
Some additional regression tests would be welcome.
See https://musescore.org/en/user/124871/blog/2016/07/10/gsoc-week-7#comment-532536

**Comment by Jojo-Schmitz** [comment]
mfest failures in tst_implodeExplode  ?

File: mtest/CMakeLists.txt [hunk 9]
```diff
@@ -173,6 +173,7 @@ subdirs (
         libmscore/element
         libmscore/exchangevoices
         libmscore/hairpin
+	libmscore/implode_explode
```
Comment (Jojo-Schmitz) [inline mtest/CMakeLists.txt]: wrong indentation

File: mtest/libmscore/CMakeLists.txt [hunk 10]
```diff
@@ -12,8 +12,7 @@
 #=============================================================================
 
 subdirs(
-      album barline beam breath chordsymbol clef clef_courtesy compat concertpitch copypaste
-          copypastesymbollist cursor dynamic earlymusic element exchangevoices hairpin instrumentchange join keysig layout links parts measure midi      midimapping note plugins repeat selectionfilter selectionrangedelete spanners split splitstaff timesig tools transpose tuplet text
+      album barline beam breath chordsymbol clef clef_courtesy compat concertpitch copypaste copypastesymbollist cursor dynamic earlymusic element exchangevoices hairpin instrumentchange join keysig layout links parts measure midi
```
Comment (Jojo-Schmitz) [inline mtest/libmscore/CMakeLists.txt]: Where's 'midimapping note plugins repeat selectionfilter selectionrangedelete spanners split splitstaff timesig transpose tuplet text' (I understood that you moved tools)

File: mtest/CMakeLists.txt (hunk 9 above)
Comment (Rockettwo) [inline mtest/CMakeLists.txt]: Why? For me it seems to be correct.

Comment (Jojo-Schmitz) [inline mtest/CMakeLists.txt]: TAB vs. Spaces? It certainly looks wrongly indented in this diff output

File: mtest/libmscore/CMakeLists.txt (hunk 10 above)
Comment (Rockettwo) [inline mtest/libmscore/CMakeLists.txt]: Hmm they should all be there and I should have added the implode_explode direction

Comment (Jojo-Schmitz) [inline mtest/libmscore/CMakeLists.txt]: nope, they are not

File: mtest/CMakeLists.txt (hunk 9 above)
Comment (Jojo-Schmitz) [inline mtest/CMakeLists.txt]: It is a TAB, should be 8(?) spaces instead to match the other lines

**Comment by Rockettwo** [comment]
I didn't update my test case which already included a slur and I have recently added the ability to handle slurs. Now it is all updated.

**Comment by lasconic** [comment]
Hey @Rockettwo, two of your PR has been merged. Now implode and explode PR need a rebase. Can you take a look?

**Comment by lasconic** [comment]
Could you rebase one more time, sorry... I will merge it next.

**Comment by Rockettwo** [comment]
I did rebase it but probably the test needs a modification due to the last changes...

**Comment by Jojo-Schmitz** [comment]
yes, seems mtest/libmscore/implode_explode/implode1-ref.mscx and mtest/libmscore/implode_explode/undoImplode01-ref.mscx need to get adjusted

**Comment by Jojo-Schmitz** [comment]
?? How did my commit got into this?

**Comment by Rockettwo** [comment]
Hmm interesting... I did a rebase again. Should work now.

**Comment by Jojo-Schmitz** [comment]
Looks good now, time to merge?

**Comment by lasconic** [comment]
I merged the other one and now this one need a rebase...

**Pull request merged**